electrum/txbatcher.py

```diff
@@ -0,0 +1,354 @@
+                self.wallet.adb.add_transaction(tx)
+                if tx.has_change():
+                    self.batch_txids.append(tx.txid())
+                    self._base_tx = tx
+                else:
+                    self.logger.info(f'starting new batch because current base tx does not have change')
+                    self.start_new_batch(tx)
+            else:
+                # most likely reason is that base_tx is not replaceable
+                # this may be the case if it has children (because we don't pay enough fees to replace them)
+                # or if we are trying to sweep unconfirmed inputs (replacement-adds-unconfirmed error)
+                self.logger.info(f'cannot broadcast tx {tx}')
+                if base_tx:
+                    self.logger.info(f'starting new batch because could not broadcast')
+                    self.start_new_batch(base_tx)
+
+    def _create_batch_tx(self, base_tx, to_sweep, to_pay, password):
+        self.logger.info(f'to_sweep: {list(to_sweep.keys())}')
+        self.logger.info(f'to_pay: {to_pay}')
+        inputs = []
+        outputs = []
+        locktime = base_tx.locktime if base_tx else None
+        # sort inputs so that txin-txout pairs are first
+        for sweep_info in sorted(to_sweep.values(), key=lambda x: not bool(x.txout)):
+            if sweep_info.cltv_abs is not None:
+                if locktime is None or locktime < sweep_info.cltv_abs:
+                    # nLockTime must be greater than or equal to the stack operand.
+                    locktime = sweep_info.cltv_abs
+            inputs.append(copy.deepcopy(sweep_info.txin))
+            if sweep_info.txout:
+                outputs.append(sweep_info.txout)
+        self.logger.info(f'locktime: {locktime}')
+        outputs += to_pay
+        inputs += self._create_inputs_from_tx_change(self._parent_tx) if self._parent_tx else []
+        tx = self.wallet.create_transaction(
+            base_tx=base_tx,
+            inputs=inputs,
+            outputs=outputs,
+            password=password,
+            locktime=locktime,
+            BIP69_sort=False,
+            merge_duplicate_outputs=False,
+        )
+        # this assert will fail if we merge duplicate outputs
+        for o in outputs: assert o in tx.outputs()
+        assert tx.is_complete()
+        return tx
+
+    def _clear_unconfirmed_sweeps(self, tx):
+        # this ensure that we can accept an input again
+        # if the spending tx has been removed from the blockchain
+        # fixme: what if there are several batches?
+        for txin in tx.inputs():
+            if txin.prevout in self._unconfirmed_sweeps:
+                self._unconfirmed_sweeps.remove(txin.prevout)
+
+    def _start_new_batch(self, tx):
+        use_change = tx and tx.has_change() and any([txout in self.batch_payments for txout in tx.outputs()])
+        self.batch_payments = self._to_pay_after(tx)
```

Comment:
does this not need the lock?
```suggestion
        with self.lock:
            self.batch_payments = self._to_pay_after(tx)
```

Reply:
done